File: code/frontend/Tyr/src/services/_utils/Either.ts
/**
 * Either type for handling success or error
 *
 * @param L - The error type
 * @param R - The success type
 */
export type Either<L, R> = Failure<L> | Success<R>;

/**
 * Success type
 *
 * @param value The value to be stored in the success type
 */
export function success<R>(value: R): Either<never, R> {
    return new Success<R>(value);
}

/**
 * Failure type
 *
 * @param error The error to be stored in the failure type
 */
export function failure<L>(error: L): Either<L, never> {
    return new Failure<L>(error);
}

/**
 * Represents Failure
 */
export class Failure<L> {
    constructor(public value: L) {}
}

/**
 * Represents Success
 */
export class Success<R> {
    constructor(public value: R) {}
}

File: code/backend/Odin/src/main/kotlin/pt/isel/odin/utils/Either.kt
package pt.isel.odin.utils

/**
 * Either type
 * @see Left
 * @see Right
 */
sealed class Either<out L, out R> {
    data class Left<out L>(val value: L) : Either<L, Nothing>()
    data class Right<out R>(val value: R) : Either<Nothing, R>()
}

/**
 * Creates a [Either.Right] type.
 * @param value The value to be stored in the [Either.Right].
 */
fun <R> success(value: R) = Either.Right(value)

/**
 * Creates a [Either.Left] type.
 * @param error The error to be stored in the [Either.Left].
 */
fun <L> failure(error: L) = Either.Left(error)

/**
 * Represents Success
 */
typealias Success<S> = Either.Right<S>

/**
 * Represents Failure
 */
typealias Failure<F> = Either.Left<F>
